File: 2684 Maximum Number of Moves in a Grid/2684.py
class Solution:
    def maxMoves(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dp = [[0]*n for _ in range(m)]

        self.ans = 0

        def valid_row(r):
            return r >= 0 and r < m

        def valid_move(r, c, next_r, next_c, next_cnt):
            return valid_row(next_r) and grid[r][c] < grid[next_r][next_c] and next_cnt > dp[next_r][next_c]

        def find(r, c):
            if c < n - 1:
                next_cnt = dp[r][c] + 1
                next_c = c + 1
                for next_r in range(r-1, r+2):
                    if valid_move(r, c, next_r, next_c, next_cnt):
                        self.ans = max(self.ans, next_cnt)
                        dp[next_r][next_c] = next_cnt
                        find(next_r, next_c)

        for r in range(m):
            find(r, 0)

        return self.ans

File: 2684 Maximum Number of Moves in a Grid/2684.cpp
class Solution {
public:
    bool validRow(vector<vector<int>>& grid, int row){
        return row >= 0 && row < grid.size();
    }
    
    void find(vector<vector<int>>& grid, vector<vector<int>>& dp, int r, int c, int& ans){
        int m = grid.size();
        int n = grid[0].size();
        
        if (c < n - 1) {
            int nextCnt = dp[r][c] + 1;
            int nextC = c + 1;
            for (int nextR = r-1; nextR <= r+1; nextR++){
                if (validRow(grid, nextR) && nextCnt > dp[nextR][nextC] && grid[r][c] < grid[nextR][nextC]){  
                    ans = max(ans, nextCnt);
                    dp[nextR][nextC] = nextCnt;
                    find(grid, dp, nextR, nextC, ans);
                }
            }   
        }
    }

    int maxMoves(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));

        int ans = 0;

        for (int r = 0; r < m; r++){
            find(grid, dp, r, 0, ans);
        }

        return ans;
    }
};
